src/dataqueue/queue.cc

```diff
@@ -0,0 +1,1221 @@
+  }
+
+  std::unique_ptr<Entry> slice(
+      uint64_t start, Maybe<uint64_t> end = Nothing<uint64_t>()) override {
+    std::shared_ptr<DataQueue> sliced = data_queue_->slice(start, end);
+    if (!sliced) return nullptr;
+
+    return std::make_unique<DataQueueEntry>(std::move(sliced));
+  }
+
+  // Returns the number of bytes represented by this Entry if it is
+  // known. Certain types of entries, such as those backed by streams
+  // might not know the size in advance and therefore cannot provide
+  // a value. In such cases, size() must return v8::Nothing<uint64_t>.
+  //
+  // If the entry is idempotent, a size should always be available.
+  Maybe<uint64_t> size() const override { return data_queue_->size(); }
+
+  // When true, multiple reads on the object must produce the exact
+  // same data or the reads will fail. Some sources of entry data,
+  // such as streams, may not be capable of preserving idempotency
+  // and therefore must not claim to be. If an entry claims to be
+  // idempotent and cannot preserve that quality, subsequent reads
+  // must fail with an error when a variance is detected.
+  bool isIdempotent() const override { return data_queue_->isIdempotent(); }
+
+  void MemoryInfo(node::MemoryTracker* tracker) const override {
+    tracker->TrackField("data_queue", data_queue_);
+  }
+
+  DataQueue& getDataQueue() { return *data_queue_; }
+
+  SET_MEMORY_INFO_NAME(DataQueueEntry);
+  SET_SELF_SIZE(DataQueueEntry);
+
+ private:
+  std::shared_ptr<DataQueue> data_queue_;
+
+  class ReaderImpl : public DataQueue::Reader {
+   public:
+    explicit ReaderImpl(std::shared_ptr<DataQueue::Reader> inner)
+        : inner_(std::move(inner)) {}
+
+    int Pull(DataQueue::Reader::Next next,
+             int options,
+             DataQueue::Vec* data,
+             size_t count,
+             size_t max_count_hint) override {
+      return inner_->Pull(
+          std::move(next), options, data, count, max_count_hint);
+    }
+
+    SET_NO_MEMORY_INFO()
+    SET_MEMORY_INFO_NAME(ReaderImpl)
+    SET_SELF_SIZE(ReaderImpl)
+
+   private:
+    std::shared_ptr<DataQueue::Reader> inner_;
+  };
+};
+
+// ============================================================================
+
+// An FdEntry reads from a file descriptor. A check is made before each read
+// to determine if the fd has changed on disc. This is a best-effort check
+// that only looks at file size, creation, and modification times. The stat
+// check is also async, so there's a natural race condition there where the
+// file could be modified between the stat and actual read calls. That's
+// a tolerable risk here. While FdEntry is considered idempotent, this race
+// means that it is indeed possible for multiple reads to return different
+// results if the file just happens to get modified.
+class FdEntry final : public EntryBase {
+  // TODO(@jasnell, @flakey5):
+  // * This should only allow reading from regular files. No directories, no
+  // pipes, etc.
```

Comment:
No pipes? This seems like an odd TODO…

Reply:
Really the requirement here is that reads must be reliably repeatable with a known size up front.

Reply:
I can see that the size requirement, if there is one, makes a difference here, but reads are not idempotent either way, because most files can also independently be modified.

Reply:
yes, per the comment for `FdEntry`, the idempotency is best-effort. We examine the size and modification time on each read to make a quick determination if the file has changed, acknowledging that yes it may have.